subsys/net/lib/coap/coap_client.c

```diff
@@ -385,6 +387,25 @@ static void report_callback_error(struct coap_client_internal_request *internal_
 	}
 }
 
+int coap_client_cancel_req(struct coap_client *client)
+{
+	for (int i = 0; i < CONFIG_COAP_CLIENT_MAX_REQUESTS; i++) {
+		if (client->requests[i].request_ongoing == true) {
+			LOG_DBG("Cancelling request %d");
+			(void)k_mutex_lock(&client->send_mutex, K_NO_WAIT);
+			coap_pending_clear(&client->requests[i].pending);
+			k_mutex_unlock(&client->send_mutex);
+			report_callback_error(&client->requests[i],
+					      COAP_RESPONSE_CODE_GATEWAY_TIMEOUT);
+			client->requests[i].request_ongoing = false;
+		}
+	}
+	atomic_clear(&coap_client_recv_active);
+	/* Wait until after zsock_poll() times out. */
+	k_sleep(K_MSEC(COAP_PERIODIC_TIMEOUT));
```

Comment:
Why does it have to sleep?
I don't really understand.

Reply:
Well, it can take up to COAP_PERIODIC_TIMEOUT milliseconds before zsock_poll() times out. Once it does, handle_poll() will be able to return to its caller and clean up. It "feels right" to allow that to happen before returning so the system is idle again.